tiledb/sm/array/uri_manager.h

```diff
@@ -0,0 +1,257 @@
+/**
+ * @file   uri_manager.h
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2017-2021 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ This file defines class URIManager.
+ */
+
+#ifndef TILEDB_URI_MANAGER_H
+#define TILEDB_URI_MANAGER_H
+
+#include "tiledb/common/status.h"
+#include "tiledb/sm/filesystem/uri.h"
+#include "tiledb/sm/storage_manager/storage_manager.h"
+
+#include <unordered_map>
+
+using namespace tiledb::common;
+
+namespace tiledb {
+namespace sm {
+
+/**
+ * Manages the various URIs inside an array directory, considering
+ * various versions of the on-disk TileDB format.
+ */
+class URIManager {
```

Comment:
Let's talk about the name of the class. Names like "manager", "helper", etc. are programming weasel words. They should be avoided in favor of descriptive names that indicate their responsibilities crisply. In addition, classes should be named for what they are, not for what they are not but aspire to be. Aspirations belong in documentation, where they may useful indicate what the class might eventually become, and how that direction may have informed the structure of the class.

In the present case, this class looks like it's neither about URIs or about management. URIs are used as identifiers for something else, in this case elements of the storage of an array. The only things this class is doing is loading them from storage. An appropriate name for what this class is today might be `StorageElementLoader`.